K8SSAND-1885 ⁃ Stabilize the unit/integration tests flakes

We recently started seeing more and more transient errors on the unit/integration tests.
Looking at the logs, it seems like the add-dc tests are responsible for a majority of these failures:

```
--- FAIL: TestK8ssandraCluster (239.38s)
    --- FAIL: TestK8ssandraCluster/AddDcToExistingCluster (54.51s)
        --- FAIL: TestK8ssandraCluster/AddDcToExistingCluster/SchemaDisagreementOnStargate (2.80s)
            add_dc_test.go:104: wait for the CassandraInitialized condition to be set
            add_dc_test.go:718: add Stargate and Reaper
            add_dc_test.go:560: check that stargate sg1 is created
            add_dc_test.go:563: update stargate sg1 status to ready
            add_dc_test.go:575: check that reaper reaper1 is created
            add_dc_test.go:578: update reaper reaper1 status to ready
            add_dc_test.go:734: add dc2 to cluster
            add_dc_test.go:759: 
                	Error Trace:	add_dc_test.go:759
                	            				add_dc_test.go:584
                	            				add_dc_test.go:223
                	Error:      	Received unexpected error:
                	            	Operation cannot be fulfilled on k8ssandraclusters.k8ssandra.io "add-dc-test": the object has been modified; please apply your changes to the latest version and try again
                	Test:       	TestK8ssandraCluster/AddDcToExistingCluster/SchemaDisagreementOnStargate
                	Messages:   	failed to add dc to K8ssandraCluster
```

This is starting to fail too often for us to re-run failed jobs now, and we should fix whatever race condition is happening here.

--- a/CHANGELOG/CHANGELOG-1.5.md
+++ b/CHANGELOG/CHANGELOG-1.5.md
@@ -14,4 +14,6 @@ Changelog for the K8ssandra Operator, new PRs should update the `unreleased` sec
 When cutting a new release, update the `unreleased` heading to the tag being generated and date, like `## vX.Y.Z - YYYY-MM-DD` and create a new placeholder section for  `unreleased` entries.
 
 ## unreleased
+
 * [FEATURE] [#599](https://github.com/k8ssandra/k8ssandra-operator/issues/600) Disable secrets management and replication with the external secrets provider
+* [TESTING] [#761](https://github.com/k8ssandra/k8ssandra-operator/issues/761) Stabilize integration and e2e tests

--- a/controllers/k8ssandra/remove_dc_test.go
+++ b/controllers/k8ssandra/remove_dc_test.go
@@ -86,7 +86,7 @@ func deleteDcWithUserKeyspacesFails(ctx context.Context, t *testing.T, f *framew
 func deleteDcWithUserKeyspacesSucceeds(ctx context.Context, t *testing.T, f *framework.Framework, kc *api.K8ssandraCluster) {
 	require := require.New(t)
 
-	replication := map[string]int{"dc1": 3}
+	replication := map[string]int{"dc1": 3, "dc2": 3}
 	updatedReplication := map[string]int{"dc1": 3}
 
 	// We need a version of the map with string values because GetKeyspaceReplication returns

--- a/test/e2e/reaper_test.go
+++ b/test/e2e/reaper_test.go
@@ -167,15 +167,6 @@ func createMultiReaper(t *testing.T, ctx context.Context, namespace string, f *f
 		username, password := retrieveCredentials(t, f, ctx, secretKey)
 		testReaperApi(t, ctx, f.DataPlaneContexts[1], DcClusterName(t, f, dc2Key), "reaper_ks", username, password)
 	})
-
-	replication := map[string]int{"dc1": 1, "dc2": 1}
-
-	t.Run("TestStargateApi[0]", func(t *testing.T) {
-		testStargateApis(t, f, ctx, f.DataPlaneContexts[0], namespace, dc1Prefix, username, password, false, replication)
-	})
-	t.Run("TestStargateApi[1]", func(t *testing.T) {
-		testStargateApis(t, f, ctx, f.DataPlaneContexts[1], namespace, dc2Prefix, username, password, false, replication)
-	})
 }
 
 func createMultiReaperWithEncryption(t *testing.T, ctx context.Context, namespace string, f *framework.E2eFramework) {

--- a/test/e2e/stargate_apis_test.go
+++ b/test/e2e/stargate_apis_test.go
@@ -4,14 +4,15 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/labels"
 	"net/http"
 	neturl "net/url"
-	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
 	"strconv"
 	"testing"
 	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
 
 	"github.com/datastax/go-cassandra-native-protocol/client"
 	"github.com/datastax/go-cassandra-native-protocol/frame"
@@ -411,11 +412,17 @@ func insertRowsGrpc(t *testing.T, grpc *grpcclient.StargateClient, nbRows int, t
 }
 
 func checkRowCountGrpc(t *testing.T, grpc *grpcclient.StargateClient, nbRows int, tableName, keyspaceName string) {
-	cql := fmt.Sprintf("SELECT id FROM %s.%s", keyspaceName, tableName)
-	response, err := grpc.ExecuteQuery(&grpcproto.Query{Cql: cql})
-	if assert.NoError(t, err, "Query failed: %s", cql) {
-		assert.Len(t, response.GetResultSet().GetRows(), nbRows, "Expected SELECT query to return %d rows", nbRows)
-	}
+	assert.Eventually(t, func() bool {
+		cql := fmt.Sprintf("SELECT id FROM %s.%s", keyspaceName, tableName)
+		response, err := grpc.ExecuteQuery(&grpcproto.Query{Cql: cql})
+		if err != nil {
+			t.Logf("Query failed: %s with error %s", cql, err)
+			return false
+		} else {
+			t.Logf("Expected %d rows, got %d", nbRows, len(response.GetResultSet().GetRows()))
+			return len(response.GetResultSet().GetRows()) == nbRows
+		}
+	}, time.Minute, time.Second*10, "failed checking count through gRPC")
 }
 
 func formatReplicationForRestApi(replication map[string]int) string {

--- a/test/e2e/suite_test.go
+++ b/test/e2e/suite_test.go
@@ -5,7 +5,6 @@ import (
 	"encoding/json"
 	"flag"
 	"fmt"
-	"github.com/k8ssandra/k8ssandra-operator/apis/config/v1beta1"
 	"net/http"
 	"net/url"
 	"os"
@@ -13,6 +12,8 @@ import (
 	"strings"
 	"testing"
 	"time"
+
+	"github.com/k8ssandra/k8ssandra-operator/apis/config/v1beta1"
 
 	reaperclient "github.com/k8ssandra/reaper-client-go/reaper"
 	"gopkg.in/resty.v1"
@@ -667,8 +668,8 @@ func cleanUp(t *testing.T, f *framework.E2eFramework, opts *e2eTestOpts) error {
 		t.Logf("failed to dump cluster info: %v", err)
 	}
 
-	timeout := 3 * time.Minute
-	interval := 10 * time.Second
+	timeout := 10 * time.Minute
+	interval := 15 * time.Second
 
 	if !opts.skipK8ssandraClusterCleanup {
 		if err := f.DeleteK8ssandraClusters(opts.sutNamespace, timeout, interval); err != nil {

